### DISCUSSION TITLE
Downloads graph at bottom of page overwhelmed by alpha versions of crate

### DISCUSSION THREAD

**Discussion by tspiteri** [rust-lang/crates.io#5454] [Feature Requests]
### Current Behavior

For example for my [*fixed* crate](https://crates.io/crates/fixed) I have seven 2.0.0-alpha versions, (six of which are yanked). The graph at the bottom of the page just shows these, so the graphs for the stable 1.* versions are not visible.

![image](https://user-images.githubusercontent.com/18604588/200825218-24c53a2e-73f7-4be0-a4ba-e0db9259609c.png)

It seems that currently the graph just shows the first five versions using semver ordering.

### Expected Behavior

I didn't expect alpha/prerelease/beta versions, most of which are yanked, to completely overwhelm the graph so that no stable versions are visible. Some options that would solve this, though I don't know if any of them is even preferable to the current semver order:
* sort by semver, but do not include yanked versions
* sort by semver, but include only the highest for any x.x.x version number, so that for example from 2.0.0-alpha.1, 2.0.0-alpha.2, 2.0.0-beta.1, 2.0.0-rc.1, only one line for 2.0.0-rc.1 would be shown
* sort by release date

**Comment by Turbo87** [comment]
I've converted this to a discussion because the feature is working as designed, which is to "show the latest X versions". I agree that this is less useful than it could be in this case, but I'd rather treat this as an improvement than an actual bug.

**Comment by cuviper** [comment]
There's a similar issue with the graph on [indexmap](https://crates.io/crates/indexmap) even with just regular releases:

![image](https://github.com/rust-lang/crates.io/assets/36186/2673ab14-3eca-4538-b4e9-d13802fbc88f)

You can see that the `2.*` uptake is decent, but there's still a whole lot of "other" that I expect are `1.*` users. In this case, I think I'd rather see "latest, plus the next X most popular", rather than just the latest X. Maybe there could be a dropdown like the stacked/unstacked choice to select different sorting modes?

**Reply by eth3lbert** [reply]
The requested enhancement sounds reasonable. I'll create a pull request for it.

I'm unsure whether keeping the latest version in other sorting options is crucial though. I think implement a dropdown menu that allows users to choose between sorting by either semver or downloads might be sufficient?

However, I'm a bit afraid about the performance of sorting by downloads for the large table in production. We'll have to monitor it and see how it goes.

POC screenshots
fixed:
| semver | downloads |
| --- | --- |
| <img width="935" alt="image" src="https://github.com/rust-lang/crates.io/assets/2889664/f1fec717-7070-4701-b90f-19928d1dcf34"> | <img width="939" alt="image" src="https://github.com/rust-lang/crates.io/assets/2889664/c7cb7044-ad92-4747-afba-1b7cfd9803e3"> |

indexmap:
| semver | downloads |
| --- | --- |
| <img width="934" alt="image" src="https://github.com/rust-lang/crates.io/assets/2889664/2f9417a8-65d9-4197-b8b0-a46aa3b0c4dd"> | <img width="934" alt="image" src="https://github.com/rust-lang/crates.io/assets/2889664/ba3d2d90-7288-4438-9c57-2c528e0cf006"> |

**Reply by cuviper** [reply]
Hmm -- with indexmap, sorting by downloads (by 90-day sum?) lost all visibility into the rise of the most recent 2.2.3. It should show up eventually, but it might take a while to smooth out. There's still a whole lot of "other" too, sadly.

**Reply by eth3lbert** [reply]
> sorting by downloads (by 90-day sum?)

Yes!

> lost all visibility into the rise of the most recent 2.2.3. It should show up eventually, but it might take a while to smooth out.

Maybe we could mitigate this by using weighted arithmetic, where recent downloads have a higher weight.

> There's still a whole lot of "other" too

I have no idea how to mitigate this issue without either increasing the number of labels or changing the visualization.

**Reply by eth3lbert** [reply]
Select candidates based on the sum of their downloads over the past 7 days:
<img width="938" alt="image" src="https://github.com/rust-lang/crates.io/assets/2889664/b2596398-f90a-468c-ab98-c151739acccc">

**Reply by cuviper** [reply]
I like that!

I just noticed in your previous "downloads" graph, the total stack is higher than in the semver view, which shouldn't happen -- I wonder if perhaps some listed versions were double-counted in "other". Your new 7-day graph doesn't have that problem.

**Reply by eth3lbert** [reply]
Oops, thanks for pointing it out. It must have been something wrong before. I just updated the graph.

**Reply by cuviper** [reply]
Great, with that update, I like both versions of the download graphs. If we have to pick just one, I guess 90 makes more sense in general, but the 7 feels more current. They'll probably be about the same for a lot of crates though.

**Reply by eth3lbert** [reply]
Yep, 7 days or any period far less than 90 days is more like recent trends.

**Reply by EdJoPaTo** [reply]
On the fixed crate the prerelease versions are still polluting the semver graph making it basically useless. What about a toggle to only show non-prerelease, non-yanked semver versions?
